File: src/python/logger.py
from datetime import datetime
import os

LOG_FILE = "logs/system.log"

# ANSI color codes for terminal output
RESET = "\033[0m"
RED = "\033[31m"
YELLOW = "\033[33m"
NORMAL = "\033[0m"

# Ensure the logs directory exists
os.makedirs(os.path.dirname(LOG_FILE), exist_ok=True)

def log(level: str, source: str, message: str):
    # Get the current timestamp
    timestamp = datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S")

    # Format the log entry
    log_entry = f"[{timestamp}] [{level}] [{source}] {message}"

    # Write the log entry to the log file
    try:
        with open(LOG_FILE, "a") as log_file:
            log_file.write(log_entry + "\n")
    except Exception as e:
        print(f"Error writing to log file: {e}")

    # Print the log entry to the console with color
    color = NORMAL
    if level == "ERROR":
        color = RED
    elif level == "WARNING":
        color = YELLOW
    print(f"{color}{log_entry}{RESET}")

File: src/cpp/logger/logger.cpp
#include "logger.hpp"
#include <fstream>
#include <iostream>
#include <filesystem>
#include <ctime>
#include <iomanip>
#include <sstream>

// ansi color codes
#define RESET   "\033[0m"
#define RED     "\033[31m"
#define YELLOW  "\033[33m"
#define NORMAL  "\033[0m"

// macro to simplify logging calls
#define LOG(level, source, message) Logger::log(level, source, message, __FILE__, __LINE__)

void Logger::log(const std::string& level, const std::string& message, const std::string& file, int line) {
    // define the log file path
    const std::string logDirPath = "logs";
    const std::string logFilePath = logDirPath + "/system.log";

    // ensure the logs directory exists
    if (!std::filesystem::exists(logDirPath)) {
        if (!std::filesystem::create_directories(logDirPath)) {
            std::cerr << "Error: Unable to create directory: " << logDirPath << std::endl;
            return;
        }
    }

    // get the current time
    std::time_t now = std::time(nullptr);
    std::tm* localTime = std::localtime(&now);
    std::ostringstream timestamp;
    timestamp << std::put_time(localTime, "%Y-%m-%d %H:%M:%S");

    // format the log message
    std::ostringstream logEntry;
    logEntry << "[" << timestamp.str() << "] "
             << "[" << level << "] "
             << "[" << file << ":" << line << "] "
             << message;

    // write the log message to the log file
    std::ofstream logFile(logFilePath, std::ios::app);
    if (logFile.is_open()) {
        logFile << logEntry.str() << std::endl;
        logFile.close();
    } else {
        std::cerr << "Error: Unable to open log file: " << logFilePath << std::endl;
    }

    // determine color based on log level
    std::string color;
    if (level == "ERROR") {
        color = RED;
    } else if (level == "WARNING") {
        color = YELLOW;
    } else {
        color = NORMAL;
    }

    // format the log message
    std::ostringstream consoleEntry;
    consoleEntry << "[" << timestamp.str() << "] "
             << "[" << level << "] "
             << "[" << file << ":" << line << "] "
             << message;

    // print the log message to the console with color
    std::cout << color << consoleEntry.str() << RESET << std::endl;
}